pkg/provider/ibmcloud/plugin/loadbalancer/lb.go

```diff
@@ -0,0 +1,488 @@
+package ibmcloud
+
+import (
+	"fmt"
+	"strings"
+	"sync"
+	"time"
+
+	logutil "github.com/docker/infrakit/pkg/log"
+	"github.com/docker/infrakit/pkg/spi/instance"
+	"github.com/docker/infrakit/pkg/spi/loadbalancer"
+	"github.com/softlayer/softlayer-go/datatypes"
+	"github.com/softlayer/softlayer-go/services"
+	"github.com/softlayer/softlayer-go/session"
+)
+
+// For logging
+var lbLogger = logutil.New("module", "ibmcloud/loadbalancer")
+
+const (
+	debugV           = logutil.V(100)
+	updateRetryCount = 30
+	updateRetryTime  = 100 * time.Millisecond
+)
+
+// Result string
+type lbResult string
+
+// LBOptions are the configuration parameters for the LB provisioner.
+type LBOptions struct {
+	uuid string
+}
+
+// ibmcloudlb captures the options for starting up the plugin.
+type ibmcloudlb struct {
+	// name
+	name string
+
+	// UUID of the IBM Cloud load balancer backing this
+	uuid string
+
+	// mutex to prevent overlaps of calls. May not be needed.
+	lock sync.Mutex
+
+	// Softlayer user name
+	softlayerUsername string
+
+	// Softlayer API Key
+	softlayerAPIKey string
+
+	// IAM endpoint
+	iamEndpoint string
+
+	// Softlayer end point URL
+	softLayerEndpointURL string
+
+	// SoftlayerXMLRPCEndpoint endpoint
+	softlayerXMLRPCEndpoint string
+
+	// retry count for API calls
+	// Unexposed in the schema at this point as they are used only during session creation for a few calls
+	// When sdk implements it we an expose them for expected behaviour
+	// https://github.com/softlayer/softlayer-go/issues/41
+	retryCount int
+
+	// Constant retry delay for API calls
+	retryDelay time.Duration
+
+	// Session
+	session *session.Session
+
+	// Loadbalancer service
+	lbService services.Network_LBaaS_LoadBalancer
+
+	// Loadbalancer Listener service
+	lbListenerService services.Network_LBaaS_Listener
+
+	// Loadbalancer Member service
+	lbMemberService services.Network_LBaaS_Member
+
+	// Loadbalancer health Monitor  service
+	lbHealthMonitorService services.Network_LBaaS_HealthMonitor
+
+	// SSL Certificate service
+	certService services.Security_Certificate
+}
+
+// NewIBMCloudLBPlugin returns a L4 loadbalancer backed by the IBM cloud
+func NewIBMCloudLBPlugin(username, apikey, lbName, lbUUID string) (loadbalancer.L4, error) {
+	lbLogger.Debug("NewIBMCloudLBPlugin", "V", debugV, "Name", lbName)
+	lb := &ibmcloudlb{
+		softlayerUsername: username,
+		softlayerAPIKey:   apikey,
+		name:              lbName,
+		uuid:              lbUUID,
+	}
+
+	// Hardcoded IBM Cloud endpoints
+	lb.iamEndpoint = "https://iam.ng.bluemix.net"
```

Comment:
It might be useful to define a few exported / public constants and not have these values buried in here.

Reply:
Done.